### ISSUE TITLE
Scheduled Sampling and Discrepancy between Validation and Testing

### ISSUE THREAD

**Issue by zhongpeixiang** [OpenNMT/OpenNMT-py#540]
This NIPS paper about [scheduled sampling](https://arxiv.org/pdf/1506.03099.pdf) showed some improvements in sequence learning brought by scheduled sampling. I have not found any features in this project yet, do you plan to implement it in the near future?

Also, the decoding process in [validation ](https://github.com/OpenNMT/OpenNMT-py/blob/3d519f9d898931f2f7a702e4265fd55adc0d4d56/onmt/Trainer.py#L209)and [training](https://github.com/OpenNMT/OpenNMT-py/blob/3d519f9d898931f2f7a702e4265fd55adc0d4d56/onmt/Trainer.py#L290) phases are the same, which **both use the true target word to predict the next word**, whereas during [inference](https://github.com/OpenNMT/OpenNMT-py/blob/3d519f9d898931f2f7a702e4265fd55adc0d4d56/onmt/translate/Translator.py#L114) (testing phase), the **decoder predicts the next word using the previously predicted word as input**. This difference makes the validation and testing results not comparable because the validation procedure is like cheating and the validation results are probably always better than testing results.

Is this validation implementation (**using target ground truth to predict next word**) a common practice in the research community? I found that in language modelling they use this kind of validation as well.

**Comment by zhongpeixiang** [comment]
Interestingly I found this [post](http://www.inference.vc/scheduled-sampling-for-rnns-scoring-rule-interpretation/), which argues scheduled sampling is not a proper objective function.

**Comment by KelleyYin** [comment]
I rebuilt this function in this system. It has a little promotion about 0.5 Bleu4 score in Chinese between English corpus . However, I don't do much adjustment for this .

**Comment by srush** [comment]
We can implement this if there is desire. The problem is that it breaks a lot of optimizations, so we haven't included it yet.

**Comment by zhongpeixiang** [comment]
@srush Yes, implementing scheduled sampling requires iterating through output sequence step by step. Do you have a working example (with dataset, code and hyper-parameters) of dialogue system based on your project?

**Comment by srush** [comment]
I don't have one currently, but I believe people have tried it for the Lua version: 

http://forum.opennmt.net/t/english-chatbot-advice/32/4

**Comment by arendu** [comment]
Seems like OP raised two issues here. 1. is scheduled sampling and 2. is whether it is ok for the decoder to condition on the true previous token each time it generates an output. It should not condition on true tokens imo.

Is there a plan to "fix" the second point? will that also break a lot of the optimizations?

**Comment by srush** [comment]
What do you propose to do instead?

**Comment by arendu** [comment]
Having thought about this some more, I want to refine my comments. 
For Perplexity conditioning on the true target token is the right thing to do, since we are measuring the model's probability mass on *that* specific sentence.(Originally I felt this was also incorrect)

However, for accuracy, conditioning on true target token seems incorrect. One option is to run the decoder with a beam-size 1 (for speed) and measure accuracy on the decoder's output.

I don't have a sense of how hard this is in terms of implementation.

**Comment by srush** [comment]
We happen to like this approach to the issue: https://aclweb.org/anthology/D16-1137

It's not currently implemented in OpenNMT, but I would love a contribution.

**Comment by coder1729** [comment]
@srush if scheduled sampling breaks optimizations due to it requiring iteration through output sequence step by step, instead of feeding together, the same thing is valid in the above paper you mentioned. Isn't it?
